[FE] context를 통한 component Id/ data 전달

YimJiYoung opened the pull request:
### 📕 Issue Number

Close #358 


### 📙 작업 내역

> 구현 내용 및 작업 했던 내역

- [x] ComponentInfoContext 생성 
- [x] 상위 컴포넌트에서 context value 받고 하위 컴포넌트에 업데이트된 value 전달하도록 하는 ContextInfoWrapper 컴포넌트 구현
- [x] 투데이, 보관함, 상세(앨범 / 아티스트/ 플레이리스트) 페이지 컴포넌트 context 적용
- [x] MainMagazineCard, ContentCardList, MagazineList, LibraryCardList, TrackRowList 컴포넌트 context 적용
- [x] componentId에 사용되는 상수 객체(page, contentType, dataType) `identifier.ts`에 정의 


### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] 코딩컨벤션을 준수하는가?
- [x] PR과 관련없는 변경사항이 없는가?
- [ ] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?
      <br/>

### 📝 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

#### context에 저장되는 정보
```
interface ComponentInfo {
    componentId: string;
    data?: {
        type: string;
        id: number;
    };
}
```
- componentId는 page_(contentType*n)_(dataId) 와 같은 형태로 저장
(예시: 투데이 페이지의 매거진 리스트의 보아 매거진 -> today_magazine_2)
- context의 정보 사용하려면 `const compInfo = useContext(ComponentInfoContext)`와 같이 이용 가능
- (추가) 헤더, 뮤직플레이어에 대한 컴포넌트 id 적용/ 매거진 상세 페이지 적용
- (추가) 상세 페이지의 경우(예. album) componentId의 page는 그냥 album으로 했는데 id도 포함한 album-(id)와 같이 지정하는 것이 좋아 보여 추후에 수정하겠습니다.
<br/><br/>

YimJiYoung committed:
[FE] Feat: component info context 정의 및 wrapper 컴포넌트 구현 #358

- component id/ 가지고 있는 data를 하위 컴포넌트에 전달하기 위한 ComponentInfoContext 정의
- compoentId를 더하고, data를 추가하여 업데이트한 value를 쉽게 전달하기 위한 ComponentInfoWrapper 컴포넌트 구현

YimJiYoung committed:
[FE] Fix: ComponentWrapper 수정 / compoentId/ data type을 위한 상수 객체 정의

- componentWrapper data 제대로 업데이트 되도록 수정
- componentId, data의 type을 위한 상수 객체 정의

YimJiYoung committed:
[FE] Feat: 투데이 페이지에서 componentInfo context 전달 #358

- 투데이 페이지에서 CoomponentInfoWrapper를 통해 componentId 각 컨텐츠의 하위 컴포넌트로 전달
- identifier 객체에 contentType 추가
- ComponentInfoContext의 초기값 설정

YimJiYoung committed:
[FE] Feat: Library Album 페이지 componentInfo context 전달 #358

    - LibraryAlbum 페이지 컴포넌트에서 CoomponentInfoWrapper를 통해 componentId 각 컨텐츠
의 하위 컴포넌트로 전달
    - LibraryCardList에서 받은 데이터의 id와 type에 따라 컴포넌트 정보 추가로 전달

YimJiYoung committed:
[FE] Fix: ComponentInfoWrapper 버그 수정 #358

- context로 받은 componentInfo 데이터 mutable하게 업데이트할 경우 componentId가 길어지고 같은 데이터가 모든 context에  공유되는 문제 발생
-> immutable 방식으로 업데이트된 value 넘기도록 수정

YimJiYoung committed:
[FE] Feat: 보관함 아티스트/믹스테입/플레이리스트/트랙 페이지 컴포넌트 context 적용 #358

- 각 보관함 페이지 컴포넌트에 componentInfoContext 적용

YimJiYoung committed:
[FE] Feat: ContentsCardList / MagazineList component context 적용 #358

- 받은 데이터로 component id/ data 설정하여 하위 컴포넌트로 전달

YimJiYoung committed:
[FE] Feat: TrackRowdList / MainMagazineList component context 적용 #358

    - 받은 데이터로 component id/ data(type, id) 설정하여 하위 컴포넌트로 전달

YimJiYoung committed:
[FE] Feat: 앨범/아티스트/플레이리스트 페이지 componentInfo context 전달 #358

- 각 상세 페이지 컴포넌트에서 CoomponentInfoWrapper를 통해 componentId 각 컨텐츠
    의 하위 컴포넌트로 전달
- page / contentType에 각 페이지/ 컨텐츠 추가

mingd1023 reviewed (approved):
수고하셨습니다~! 복잡할 줄 알았는데 직관적이고 깔끔하게 잘 짜셨네요!!

mingd1023 commented on client/constants/identifier.ts:
```diff
@@ -0,0 +1,37 @@
+export const page = {
+    today: 'today',
+    libraryAlbum: 'library-album',
```
이렇게 constants 따로 뺀거 좋네요! 다른 부분도 미리 이렇게 짰으면 좋았을텐데ㅠㅠ

YimJiYoung committed:
Merge branch 'dev' into feat/component-info-context

YimJiYoung committed:
[FE] Feat: musicPlayer, headerSideBar 컴포넌트 id 적용 #358

- 모든 페이지에 공통적으로 있는 컴포넌트는 page와 다른 componentType으로 id 지정
- musicPlayer 내의 playcontroller, up next, playtrackcard 컴포넌트 id, data 적용

YimJiYoung committed:
[FE] Feat: 매거진 상세 페이지 컴포넌트 id 적용 #358

- 매거진 상세 페이지 component id / data 적용

YimJiYoung committed:
[FE] Fix: 상세 페이지 컴포넌트 id 수정 #358

- 해당 데이터 id를 컴포넌트 id에 포함하도록 수정

YimJiYoung committed:
[FE] Fix: 상세 페이지 context에 data 포함하도록 수정 #358

- 상세 페이지 componentinfo context에 data 포함하도록 수정
